src/tox/tox_env/api.py

```diff
@@ -169,6 +169,11 @@ def setup(self) -> None:
             self._handle_env_tmp_dir()
 
     def ensure_setup(self, recreate: bool = False) -> None:
+        platform_str: str = self.conf["platform"]
```

Comment:
I'd prefer to keep this check behind the active property. So please restore the property; and ideally, I think we'll use the property to not schedule the tox environments at all here https://github.com/tox-dev/tox/blob/rewrite/src/tox/session/cmd/run/common.py#L267, but instead set skip status for it.

Reply:
Hm, I probably can prevent to schedule the tox environment when I do the regex check inside `ToxEnv.active`.

But `active` returns a boolean, so I need to apply the same regex check later again, when I need to assemble a proper message a la "plat: skipped environment because platform linux does not match win32".

Hm, would I then trigger the message in `runcommands`?

https://github.com/tox-dev/tox/blob/41fa387a8c3e139ae7f953227dcbf57bede703a9/src/tox/session/cmd/run/single.py#L36-L45

I guess I need to spend more time with this :-)

Any additional tips appreciated.

Reply:
I think the best option is to make ToxEnvRunResult skipped an optional string, it's `None` if it's not skipped, but if it's skipped can be an arbitrary message. It also implies that active probably should be renamed to skip_reason that also returns an optional string.

Reply:
On second thought you don't need the property, can keep it here, just move it out to a method.

Reply:
Thanks! I hope to find some time on the weekend. Really enjoying this!

Reply:
I moved the regex check (platform) into its own method, but kept it in the `ensure_setup` method.

I then had a look around where `ensure_setup` is called - and there are 4 callers, but only the `_evaluate` method catches the `Skip` exception, so I think it would be better to call the newly created `check_platform` method just above the `tox_env.ensure_setup(recreate=recreate)` call...

https://github.com/tox-dev/tox/blob/34ef0e58d044782274a8a1fac6c43ad35994bb39/src/tox/session/cmd/run/single.py#L34-L45

ie...

```
    try:
        try:
            tox_env.check_platform()
            tox_env.ensure_setup(recreate=recreate)
            code, outcomes = run_commands(tox_env, no_test)
```

What do you think?

Thanks for taking the time to guide me through his PR!

Reply:
Why do we need to catch it? Can't it go under the same logic that already catches missing interpreter logic?

Reply:
Currently, the platform check is done via `check_platform` inside the `ensure_setup` method...

https://github.com/tox-dev/tox/blob/5bedb68c29b8ae54262c3770dab65bd68eb38fdb/src/tox/tox_env/api.py#L171-L185

In single.py -> _evaluate (which calls `ensure_setup`) this works already perfectly... if the the platform does not match, the `Skip` exception gets raised, and then caught and a message is printed... (analguous to not found interpreters) - all good!

https://github.com/tox-dev/tox/blob/34ef0e58d044782274a8a1fac6c43ad35994bb39/src/tox/session/cmd/run/single.py#L34-L45

But the `ensure_setup` method has also a couple of other callers.... like

https://github.com/tox-dev/tox/blob/fd915cc8481265da415b039f8464a2437737060c/src/tox/tox_env/python/virtual_env/package/api.py#L103-L108

and

https://github.com/tox-dev/tox/blob/fd915cc8481265da415b039f8464a2437737060c/src/tox/tox_env/python/virtual_env/package/api.py#L114-L120

and

https://github.com/tox-dev/tox/blob/ec50a6f2ccdf83a070d0037d7e685aa3c843821e/src/tox/provision.py#L96-L115

If there the `raise Skip` from within `check_platform` within `ensure_setup` gets triggered, nobody's there to handle it.

That's why I'd prefer to call `check_platform` directly in the try block of `_evaluate` - as there is the only place the `Skip` exception gets properly handled.

like this...
```
def _evaluate(tox_env: RunToxEnv, recreate: bool, no_test: bool) -> Tuple[bool, int, List[Outcome]]:
    skipped = False
    code: int = 0
    outcomes: List[Outcome] = []
    try:
        try:
            tox_env.check_platform()  # may raise Skip exception, but that's ok as it gets handled right below
            tox_env.ensure_setup(recreate=recreate)
            code, outcomes = run_commands(tox_env, no_test)
        except Skip as exception:
            LOGGER.warning("skipped environment because %s", exception)
            skipped = True
```

Reply:
Platform match is only enabled for the run environment. Neither packaging nor provisioning environments can't set that, so we're safe.

Reply:
Very well.

So, what's missing?

Rebase to please codecov?

Squash locally and forcepush?

Reply:
We need to ensure (add tests for it) that my statement above holds 😊

Reply:
So, I rebased.

As I am not yet familiar with the other packages, I then had a look at the coverage report, which tests call the packaging and provisioning parts, took two as a template, tried to strip them down, so I can just prove your statement from above... worked out fine on my Ubuntu laptop, but now it fails in CI, but only on Ubuntu / Python 3.8 and Windows / Python 3.7

The failures on above platform combinations seem odd.

The one error comes from `test_provision_requires_ok` which I took as a template. I can't see what's the problem. Maybe an isolation issue? Is it safe to use `pypi_index_self.use(monkeypatch)` in two tests?  I am not familiar with the monkeypatch fixture - do we need to clean up later manually?

Now, this is interesting... ` ERROR at setup of test_provisioning_platform_check_does_not_raise_unhandled_skip_exception` - so it is not the test, it is already failing in the setup :-/

I could def. need a helping hand.

Reply:
Maybe it is the session scope of the fixture which is causing the problem.

```
@pytest.fixture(scope="session")
def pypi_index_self(pypi_server: IndexServer, tox_wheels: List[Path], demo_pkg_inline_wheel: Path) -> Index:
```